### PR TITLE
Rewrite MutableObjectModelBinder

### PR DESCRIPTION
**Pull request by rynowak**
A rewrite focused on simplifying extensibility points and reducing
allocations.

/cc @dougbu

Using 3000 requests to BigModelBinding scenario for comparison

**Before**
image

**After**
image

Knocks off about 128mb (16%) of allocated bytes from this scenario. There's really a mixed bag of improvements here, mostly it comes from eliminating async state machines and intermediate state.

This changeset is really focused on simplifying the code and making the extensibility points more useful. This is one of those joyous occasions where simplicity and perf are out on a dinner date :cupid:

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultBindingMetadataProvider.cs [hunk 1]
```diff
@@ -71,7 +71,15 @@ public void CreateBindingMetadata(BindingMetadataProviderContext context)
...
+            {
+                context.BindingMetadata.PropertyBindingPredicateProvider = null;
+            }
+            else if (predicateProviders.Length == 1)
+            {
+                context.BindingMetadata.PropertyBindingPredicateProvider = predicateProviders[0];
+            }
+            else
             {
                 context.BindingMetadata.PropertyBindingPredicateProvider = new CompositePredicateProvider(
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultBindingMetadataProvider.cs]: This isn't really a perf-focused change, it makes testing this feature easier.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 2]
```diff
@@ -32,34 +32,93 @@ public Task BindModelAsync(ModelBindingContext bindingContext)
                 return TaskCache.CompletedTask;
             }
 
-            var mutableObjectBinderContext = new MutableObjectBinderContext()
-            {
-                ModelBindingContext = bindingContext,
-                PropertyMetadata = GetMetadataForProperties(bindingContext).ToArray(),
-            };
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Nuked from orbit :bomb:

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 3]
```diff
@@ -32,34 +32,93 @@ public Task BindModelAsync(ModelBindingContext bindingContext)
...
-
-            if (!(CanCreateModel(mutableObjectBinderContext)))
+            if (!(CanCreateModel(bindingContext)))
             {
                 return TaskCache.CompletedTask;
             }
 
-            return BindModelCoreAsync(bindingContext, mutableObjectBinderContext);
+            // Perf: separated to avoid allocating a state machine when we don't
+            // need to go async.
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Adding comment for these things from now on, I've seen people try to remove them

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 4]
```diff
@@ -32,34 +32,93 @@ public Task BindModelAsync(ModelBindingContext bindingContext)
...
+            }
 
-            bindingContext.Result = ModelBindingResult.Success(bindingContext.ModelName, model);
+            bindingContext.Result = ModelBindingResult.Success(bindingContext.ModelName, bindingContext.Model);
+        }
+
+        protected virtual bool CanBindProperty(ModelBindingContext bindingContext, ModelMetadata propertyMetadata)
+        {
+            var modelMetadataPredicate = bindingContext.ModelMetadata.PropertyBindingPredicateProvider?.PropertyFilter;
+            if (modelMetadataPredicate?.Invoke(bindingContext, propertyMetadata.PropertyName) == false)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: I want to do a follow up change to remove `bindingContext` from this API (the predicate). It's not useful for anything we care about.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 5]
```diff
@@ -32,34 +32,93 @@ public Task BindModelAsync(ModelBindingContext bindingContext)
...
+                    var message = property.ModelBindingMessageProvider.MissingBindRequiredValueAccessor(fieldName);
+                    bindingContext.ModelState.TryAddModelError(modelName, message);
+                }
+            }
 
-            bindingContext.Result = ModelBindingResult.Success(bindingContext.ModelName, model);
+            bindingContext.Result = ModelBindingResult.Success(bindingContext.ModelName, bindingContext.Model);
+        }
+
+        protected virtual bool CanBindProperty(ModelBindingContext bindingContext, ModelMetadata propertyMetadata)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Will add docs. You can override this to determine what properties in your model type are eligible for binding.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 6]
```diff
@@ -78,9 +137,13 @@ protected virtual bool CanUpdateProperty(ModelMetadata propertyMetadata)
             return CanUpdatePropertyInternal(propertyMetadata);
         }
 
-        internal bool CanCreateModel(MutableObjectBinderContext context)
+        protected virtual Task BindProperty(ModelBindingContext bindingContext)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Will add docs. You can override this to do something special for a specific property on your model.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 7]
```diff
@@ -108,14 +171,8 @@ internal bool CanCreateModel(MutableObjectBinderContext context)
                 return true;
             }
 
-            // 2. If it is top level object and there are no properties to bind
-            if (isTopLevelObject && context.PropertyMetadata != null && context.PropertyMetadata.Count == 0)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: This check isn't needed anymore.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 8]
```diff
@@ -157,9 +214,17 @@ private bool CanValueBindAnyModelProperties(MutableObjectBinderContext context)
             // create the model and try to bind it. OR if ALL properties of the model have a greedy source,
             // then we go ahead and create it.
             //
+            var hasBindableProperty = false;
             var isAnyPropertyEnabledForValueProviderBasedBinding = false;
-            foreach (var propertyMetadata in context.PropertyMetadata)
+            foreach (var propertyMetadata in bindingContext.ModelMetadata.Properties)
             {
+                if (!CanBindProperty(bindingContext, propertyMetadata))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: `CanBindProperty` is cheap to execute, no real reason to cache the result of it like the previous code did.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 9]
```diff
@@ -358,104 +374,6 @@ protected virtual object GetModel(ModelBindingContext bindingContext)
...
-            }
-
-            var validationInfo = GetPropertyValidationInfo(bindingContext);
-            var newPropertyFilter = GetPropertyFilter();
-            return bindingContext.ModelMetadata.Properties
-                                 .Where(propertyMetadata =>
-                                    newPropertyFilter(bindingContext, propertyMetadata.PropertyName) &&
-                                    (validationInfo.RequiredProperties.Contains(propertyMetadata.PropertyName) ||
-                                    !validationInfo.SkipProperties.Contains(propertyMetadata.PropertyName)) &&
-                                    CanUpdateProperty(propertyMetadata));
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: This was rolled into `CanBindProperty`. Waaaaaaaaaaaaaay simpler now

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 10]
```diff
@@ -358,104 +374,6 @@ protected virtual object GetModel(ModelBindingContext bindingContext)
...
-        }
-
-        private static Func<ModelBindingContext, string, bool> GetPropertyFilter()
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: This was rolled into CanBindProperty. Waaaaaaaaaaaaaay simpler now

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 11]
```diff
@@ -358,104 +374,6 @@ protected virtual object GetModel(ModelBindingContext bindingContext)
...
-            {
-                var modelMetadataPredicate = context.ModelMetadata.PropertyBindingPredicateProvider?.PropertyFilter;
-
-                return
-                    (context.PropertyFilter == null || context.PropertyFilter(context, propertyName)) &&
-                    (modelMetadataPredicate == null || modelMetadataPredicate(context, propertyName));
-            };
-        }
-
-        internal static PropertyValidationInfo GetPropertyValidationInfo(ModelBindingContext bindingContext)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: This was rolled into CanBindProperty. Waaaaaaaaaaaaaay simpler now

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 12]
```diff
@@ -358,104 +374,6 @@ protected virtual object GetModel(ModelBindingContext bindingContext)
...
-                var propertyName = propertyMetadata.PropertyName;
-
-                if (!propertyMetadata.IsBindingAllowed)
-                {
-                    // Nothing to do here if binding is not allowed.
-                    validationInfo.SkipProperties.Add(propertyName);
-                    continue;
-                }
-
-                if (propertyMetadata.IsBindingRequired)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: This is just handled in the main loop.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 13]
```diff
@@ -467,7 +385,6 @@ internal static PropertyValidationInfo GetPropertyValidationInfo(ModelBindingCon
         /// <remarks>Should succeed in all cases that <see cref="CanUpdateProperty"/> returns <c>true</c>.</remarks>
         protected virtual void SetProperty(
             ModelBindingContext bindingContext,
-            ModelMetadata metadata,
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Unused

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/MutableObjectModelBinderTest.cs [hunk 14]
```diff
@@ -19,6 +21,8 @@ namespace Microsoft.AspNetCore.Mvc.ModelBinding
 {
     public class MutableObjectModelBinderTest
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/MutableObjectModelBinderTest.cs]: Removed tests here that were targeted at features that have been removed. We had some tests that verify that things **don't** happen that are old features long gone from the code.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/MutableObjectModelBinderTest.cs [hunk 15]
```diff
@@ -604,169 +514,199 @@ public void GetModel_ModelIsNull_CallsCreateModel()
...
-            {
-                nameof(Person.DateOfBirth),
-                nameof(Person.DateOfDeath),
-                nameof(Person.ValueTypeRequired),
-                nameof(Person.ValueTypeRequiredWithDefaultValue),
-                nameof(Person.FirstName),
-                nameof(Person.LastName),
-                nameof(Person.PropertyWithDefaultValue),
-                nameof(Person.PropertyWithInitializedValue),
-                nameof(Person.PropertyWithInitializedValueAndDefault),
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/MutableObjectModelBinderTest.cs]: This is duplication, removed it.

File: test/Microsoft.AspNetCore.Mvc.IntegrationTests/MutableObjectModelBinderIntegrationTest.cs [hunk 16]
```diff
@@ -1400,7 +1400,7 @@ private class Person10
             // Arrange
             var metadataProvider = new TestModelMetadataProvider();
             metadataProvider
-                .ForType(typeof(Order10))
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.IntegrationTests/MutableObjectModelBinderIntegrationTest.cs]: This was a test bug masking a product bug.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 17]
```diff
@@ -358,104 +374,6 @@ protected virtual object GetModel(ModelBindingContext bindingContext)
...
-
-            foreach (var missingRequiredProperty in validationInfo.RequiredProperties)
-            {
-                var propertyMetadata = metadata.Properties[missingRequiredProperty];
-                var propertyName = propertyMetadata.BinderModelName ?? missingRequiredProperty;
-                var modelStateKey = ModelNames.CreatePropertyModelName(bindingContext.ModelName, propertyName);
-
-                bindingContext.ModelState.TryAddModelError(
-                    modelStateKey,
-                    bindingContext.ModelMetadata.ModelBindingMessageProvider.MissingBindRequiredValueAccessor(
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: This was a bug that I fixed. It should use `propertyMetadata.ModelBindingMessageProvider`

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 18]
```diff
@@ -157,9 +214,17 @@ private bool CanValueBindAnyModelProperties(MutableObjectBinderContext context)
...
             var isAnyPropertyEnabledForValueProviderBasedBinding = false;
-            foreach (var propertyMetadata in context.PropertyMetadata)
+            foreach (var propertyMetadata in bindingContext.ModelMetadata.Properties)
             {
+                if (!CanBindProperty(bindingContext, propertyMetadata))
+                {
+                    continue;
+                }
+
+                hasBindableProperty |= true;
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: `hasBindableProperty = true;` (shouldn't need to or it)

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs (hunk 13 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Remove the doc comment too.

**Comment by pranavkm** [comment]
:shipit: after you've added docs.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs (hunk 4 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Except that `PropertyName` isn't that useful on its own and the `IPropertyBindingPredicateProvider` doesn't have the metadata. Are you going to instead pass `ModelMetadata` for the property?

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs (hunk 7 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Why not? `CanValueBindAnyModelProperties()` will return `false` when there are no properties to bind. Did you intend to remove check at the top of that method?

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 19]
```diff
@@ -32,34 +32,93 @@ public Task BindModelAsync(ModelBindingContext bindingContext)
...
+            bindingContext.Model = GetModel(bindingContext);
+
+            foreach (var property in bindingContext.ModelMetadata.Properties)
+            {
+                if (!CanBindProperty(bindingContext, property))
+                {
+                    continue;
+                }
+
+                // ModelBindingContext.Model property values may be non-null when invoked via TryUpdateModel(). Pass
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Huh? `bindingContext.Model` will always be non-`null` given the `GetModel()` call just a few lines up.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 20]
```diff
@@ -32,34 +32,93 @@ public Task BindModelAsync(ModelBindingContext bindingContext)
...
             }
 
-            var mutableObjectBinderContext = new MutableObjectBinderContext()
-            {
-                ModelBindingContext = bindingContext,
-                PropertyMetadata = GetMetadataForProperties(bindingContext).ToArray(),
-            };
-
-            if (!(CanCreateModel(mutableObjectBinderContext)))
+            if (!(CanCreateModel(bindingContext)))
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Minor: Reduced the (parentheses).

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs (hunk 7 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Ah, got it. Previous condition was hidden -- have already checked `isTopLevelObject`.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 21]
```diff
@@ -88,7 +151,7 @@ internal bool CanCreateModel(MutableObjectBinderContext context)
             // recursion.
             //
             // First, we want to make sure this object is allowed to come from a value provider source as this binder
-            // will always include value provider data. For instance if the model is marked with [FromBody], then we
+            // will only include value provider data. For instance if the model is marked with [FromBody], then we
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: While fixing this, have similar issues in the page-long comment at [lines 192-215](https://github.com/aspnet/Mvc/blob/momb/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs#L192-L215) e.g. `may include ALL value providers`.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 22]
```diff
@@ -186,12 +251,10 @@ private bool CanValueBindAnyModelProperties(MutableObjectBinderContext context)
...
             }
 
-            if (!isAnyPropertyEnabledForValueProviderBasedBinding)
+            if (hasBindableProperty && !isAnyPropertyEnabledForValueProviderBasedBinding)
             {
-                // Either there are no properties or all the properties are marked as
-                // a non value provider based marker.
-                // This would be the case when the model has all its properties annotated with
-                // a IBinderMetadata. We want to be able to create such a model.
+                // All the properties are marked as a non value provider based marker like [FromHeader] or
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Confusing me here. Do you mean "are marked _with_ a non value provider based marker"? Something else?

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 23]
```diff
@@ -32,34 +32,93 @@ public Task BindModelAsync(ModelBindingContext bindingContext)
...
+                // instances or overwrite inner properties that are not bound. No need for this with simple values
+                // because they will be overwritten if binding succeeds. Arrays are never reused because they cannot
+                // be resized.
+                object propertyModel = null;
+                if (property.PropertyGetter != null &&
+                    property.IsComplexType &&
+                    !property.ModelType.IsArray)
+                {
+                    propertyModel = property.PropertyGetter(bindingContext.Model);
+                }
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Won't `CollectionModelBinder` (and similar) now always do the right thing, meaning `AddToProperty()` and `CallPropertyAddRange()` are now redundant and can be removed?

**Comment by dougbu** [comment]
:watch: for doc comments and a few issues described above.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Yep.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Also, all of the implementations of this **only** look at the property name.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs (hunk 23 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Taking a note to try this. Will probably be something I do after resolving other feedback.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs (hunk 21 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Changed ALL to any. If you want any more specific cleanup on the parts of the code that aren't changing I suggest you just do it and check it in :trollface:

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs (hunk 22 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Yep.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs (hunk 19 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: This comment is about the property values. It clearly says property values.

**Comment by rynowak** [comment]
@dougbu - updated. Currently going through the experiment of removing the collection code.

**Comment by rynowak** [comment]
> 354 additions, 802 deletions not shown because the diff is too large. Please use a local Git client to view these changes.

github says the diff on the tests is too high.

**Comment by dougbu** [comment]
> github says the diff on the tests is too high

Don't worry about it -- just the GitHub UI. The changes are visible in the commit(s) which is why I commented there in previous iteration.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 24]
```diff
@@ -2,8 +2,6 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Collections.Generic;
-using System.Linq;
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: :clap:

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 25]
```diff
@@ -15,9 +13,6 @@ namespace Microsoft.AspNetCore.Mvc.ModelBinding
     /// </summary>
     public class MutableObjectModelBinder : IModelBinder
     {
-        private static readonly MethodInfo CallPropertyAddRangeOpenGenericMethod =
-            typeof(MutableObjectModelBinder).GetTypeInfo().GetDeclaredMethod(nameof(CallPropertyAddRange));
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Nice little :bomb:

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Yeah, I'm super jacked about this change. Always hated we had to duplicate that code.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs (hunk 19 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: Reads as "`Model` property value(s)". In any case `TryUpdateModel()` is no longer a special case for this code.

**Comment by dougbu** [comment]
:watch: just for remaining unit test fallout.

**Comment by dougbu** [comment]
:shipit: w/ comment change discussed offline.

**Comment by rynowak** [comment]
Will merge once dev is open for checkins

**Pull request closed without merging**